src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -735,7 +555,17 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                     // DecryptData will decrypt in-place and modify these to point to the actual decrypted data, which may be smaller.
                     _decryptedBytesOffset = _internalOffset;
                     _decryptedBytesCount = payloadBytes;
-                    SecurityStatusPal status = DecryptData();
+
+
+                    SecurityStatusPal status;
+                    lock (_handshakeLock)
+                    {
+                        status = DecryptData();
+                        if (status.ErrorCode == SecurityStatusPalErrorCode.Renegotiate)
+                        {
+                            _handshakeWaiter = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
I must be missing it... where is _handshakeWaiter completed? i.e. where is {Try}SetResult called?

Reply:
It seems like I broke it in the final cleanup ;(

This is set in ReplyOnReAuthenticationAsync() now.

The sequence is:

1.  we try to decrypt under lock. 
2. if renegotiation is requested we create TaskCompletionSource still under lock. 
3. that gets cleaned up when a renegotiation is finished. 

4. On write side, we either win and there is no contention and waiting.
5. We get there when a renegotiation is pending. We wait (possibly in loop) until renegotiation is completed as there is no guarantee there is only one. However, in n practice, that is the case so we should be able to finish write/encrypt  as soon as TaskCompletionSource is triggered.
6. Thre are weird cases that renegotiation in TLS1.3 can finish without IO e.g. we get a frame to update session. That needs to be processed but it can finish very quickly. In that case, _handshakeWaiter should be null even if Encrypt failed with TryAgin under lock.